docs/Customization/Firrtl-Transforms.rst

```diff
@@ -0,0 +1,73 @@
+
+.. literalinclude:: ../../common.mk
+    :language: make
+    :start-after: DOC include start: FirrtlCompiler
+    :end-before: DOC include end: FirrtlCompiler
+
+If you look inside of the `tools/barstools/tapeout/src/main/scala/transforms/Generate.scala <https://github.com/ucb-bar/barstools/blob/master/tapeout/src/main/scala/transforms/Generate.scala>`__ file,
+you can see that FIRRTL invoked twice, once for the "Top" and once for the "Harness". If you want to add transforms to just modify the DUT, you can add them to ``topTransforms``.
+Otherwise, if you want to add transforms to just modify the test harness, you can add them to ``harnessTransforms``.
+
+Examples of transforms
+----------------------
+
+There are multiple examples of transforms that you can apply and are spread across the FIRRTL ecosystem.
+Within FIRRTL there is a default set of supported transforms located in https://github.com/freechipsproject/firrtl/tree/master/src/main/scala/firrtl/transforms.
+This includes transforms that can flatten modules (``Flatten``), group modules together (``GroupAndDedup``), and more.
+
+Transforms can be standalone or can take annotations as input. Annotations are used to pass information between FIRRTL transforms. This includes information on
+what modules to flatten, group, and more. Annotations can be added to the code by
+adding them to your Chisel source or by creating the ``.json`` file and adding it to the FIRRTL compiler (note: adding to the Chisel source will add the ``json`` snippet into the build system for you).
+**The recommended way to annotate something is to do it in the Chisel source**.
+
+Here is an example of adding an annotation within the Chisel source. This example is taken from the
+``firechip`` project and uses an annotation to mark BOOM's register file for optimization:
+
+.. literalinclude:: ../../generators/firechip/src/main/scala/TargetMixins.scala
+    :language: make
+    :start-after: DOC include start: ChiselAnnotation
+    :end-before: DOC include end: ChiselAnnotation
+
+Here is an example of grouping a series of modules by specifying the ``.json`` file.
+
+.. code-block:: json
+
+   [
+       {
+           "class": "firrtl.transforms.GroupAnnotation",
+           "components": [
+               "BeagleChipTop.BeagleChipTop.instance1",
+               "BeagleChipTop.BeagleChipTop.instance2",
+           ],
+           "newModule": "NewModule",
+           "newInstance": "newModInst"
+       }
+   ]
+
+In this case, the specific syntax depends on the type of annotation. The best way to figure out
+what the ``json`` file should contain is to first try to annotate in the Chisel
+source and then see what the ``json`` output gives and copy that format.
+
+Once ``yourAnnoFile.json`` is created then you can add ``-faf yourAnnoFile.json`` to the FIRRTL compiler invocation in ``common.mk``.
```

Comment:
Do we want to support this more cleanly in the build system?

Reply:
To me, it seems like you all want the main form of adding annotations to be in Chisel. If thats the case, the Makefile doesn't necessarily need to support this...

The easiest way to support this is to add a `FIRRTL_COMPILER_FLAGS` variable that is empty and could be overridden.

Reply:
This can be addressed in a later PR (after the tutorial). You're invited to open an issue